### PR TITLE
API: Make describe changes backwards compatible

### LINKED ISSUE #33903: API: Revert changes to describe
https://github.com/pandas-dev/pandas/pull/30209/ has some changes to describe for datelike data. I think the changes count as API breaking and should wait until a 2.0, possibly with a deprecation warning.

cc @david-cortes, @jreback.

master:

```python
In [1]: import pandas as pd

In [2]: pd.Series(pd.date_range('2000', periods=10)).describe()
Out[2]:
count                     10
mean     2000-01-05 12:00:00
min      2000-01-01 00:00:00
25%      2000-01-03 06:00:00
50%      2000-01-05 12:00:00
75%      2000-01-07 18:00:00
max      2000-01-10 00:00:00
dtype: object

In [3]: pd.__version__
Out[3]: '1.1.0.dev0+1417.g12c9c626d'
```

1.0.2:

```python
In [1]: import pandas as pd

In [2]: pd.Series(pd.date_range('2000', periods=10)).describe()
Out[2]:
count                      10
unique                     10
top       2000-01-08 00:00:00
freq                        1
first     2000-01-01 00:00:00
last      2000-01-10 00:00:00
dtype: object

In [3]: pd.__version__
Out[3]: '1.0.2'
```

### PR DESCRIPTION
**Pull request by TomAugspurger**
Adds the new behavior as a feature flag / deprecation.

Closes https://github.com/pandas-dev/pandas/issues/33903

(Do we have a list of issues for deprecations introduced in 1.x?)

### CODE REVIEW

**Comment by TomAugspurger** [comment]
cc @david-cortes.

File: pandas/core/generic.py [hunk 1]
```diff
@@ -10073,8 +10081,37 @@ def describe_categorical_1d(data):
...
+                        "`.describe` is deprecated and will be removed in a future "
+                        "version of pandas. Specify `datetime_is_numeric=True` to "
+                        "silence this warning and adopt the future behavior now.",
+                        FutureWarning,
+                        stacklevel=stacklevel,
+                    )
+                    tz = data.dt.tz
+                    asint = data.dropna().values.view("i8")
+                    top = Timestamp(top)
+                    if top.tzinfo is not None and tz is not None:
```
Comment (jreback) [inline pandas/core/generic.py:10100]: wow you need all this? includingthe localization?

File: pandas/tests/frame/methods/test_describe.py [hunk 2]
```diff
@@ -267,7 +267,40 @@ def test_describe_tz_values(self, tz_naive_fixture):
             },
             index=["count", "mean", "min", "25%", "50%", "75%", "max", "std"],
         )
-        result = df.describe(include="all")
+        result = df.describe(include="all", datetime_is_numeric=True)
+        tm.assert_frame_equal(result, expected)
+
+        s1_ = s1.describe()
```
Comment (jreback) [inline pandas/tests/frame/methods/test_describe.py:273]: can you make as a separate test

File: pandas/tests/series/methods/test_describe.py [hunk 3]
```diff
@@ -98,3 +98,19 @@ def test_describe_with_tz(self, tz_naive_fixture):
             index=["count", "mean", "min", "25%", "50%", "75%", "max"],
         )
         tm.assert_series_equal(result, expected)
+
+        with tm.assert_produces_warning(FutureWarning):
```
Comment (jreback) [inline pandas/tests/series/methods/test_describe.py:102]: same

**Review by jreback — CHANGES_REQUESTED** [review]
can you add to the deprecation removal list as well for 2.0

**Comment by jorisvandenbossche** [comment]
> (Do we have a list of issues for deprecations introduced in 1.x?)

https://github.com/pandas-dev/pandas/issues/30228

File: pandas/core/generic.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/core/generic.py:10100]: This is just a copy-paste of the old implementation, I didn't look closely at it, given that we want to change the default behavior in 2.x and (I think) remove the ability to treat datetimes as non-numeric after that.

**Comment by jorisvandenbossche** [comment]
BTW, some other remarks on the new behaviour: this should also be enabled in DataFrame.describe by default? 
Now (on master), a datetime is not included by default. Which made sense before, as it was regarded as a categorical and not numeric columns. But with `datetime_as_numeric`, it should be included by default?

Also, I would include "std" as well for datetimes, but with a NaN entry. That makes that the keys are always all the same for all numeric types, regardless of the exact dtype (and that also ensures the ordering is preserved when when doing describe on a dataframe with both numeric and datetime columns)

**Comment by TomAugspurger** [comment]
I was surprised to see that master didn't treat datetimes as numeric in `describe`. But I don't know how high-priority fixing that is (though a keyword calling `datetime_is_numeric` does bump the priority).

**Comment by jorisvandenbossche** [comment]
I think we should fix it, because if we add a keyword to enable the future behaviour, we should ensure we have the proper future behaviour we want (which is now not yet the case, IMO). Unless we disallow the keyword for dataframe .. 
(now, doesn't necessarily need to happen in this PR though)

**Comment by TomAugspurger** [comment]
Fixing that will get a bit messy, since it overlaps with the `include` and `exclude` keywords, and it's unclear how timedelta & period should behave (are they also numeric-like? Should the single keyword `datetime_is_numeric` control all of those? If so, should it be renamed?)

**Comment by WillAyd** [comment]
In the spirit of practicality over purity do we really need to do this? Outside of the Dask test do we expect end users would be relying on the old behavior?

**Comment by jorisvandenbossche** [comment]
> it's unclear how timedelta & period should behave (are they also numeric-like? Should the single keyword datetime_is_numeric control all of those? If so, should it be renamed?)

For timedelta that is clear, I think, since that also returns the numeric describe output, while period does not?

**Comment by jorisvandenbossche** [comment]
It seems timedelta is alraedy included by default, on released version:

```
In [7]: pd.DataFrame({'a': pd.timedelta_range("2012", periods=3), 'b': [1, 2, 3]}).describe()                                                                                                                      
Out[7]: 
                            a    b
count                       3  3.0
mean   1 days 00:00:00.000002  2.0
std           1 days 00:00:00  1.0
min    0 days 00:00:00.000002  1.0
25%    0 days 12:00:00.000002  1.5
50%    1 days 00:00:00.000002  2.0
75%    1 days 12:00:00.000002  2.5
max    2 days 00:00:00.000002  3.0
```

**Comment by TomAugspurger** [comment]
> In the spirit of practicality over purity do we really need to do this? Outside of the Dask test do we expect end users would be relying on the old behavior?

Hard to say. It's not hard to construct code that relies on the old behavior though.

**Comment by WillAyd** [comment]
My point is that this seems like a lot of churn for potentially negligible value add. I can't think of a pipeline where the old behavior is actually useful, so I don't think worth adding a keyword that we subsequently plan to deprecate for the sake of maintaining compat

**Comment by jorisvandenbossche** [comment]
There are a lot of things in pandas that I personally don't find particularly useful, but there are probably still a lot of people using those things. So I also find it hard to say whether it is important here. But it also doesn't seem difficult to actually do it with a deprecation.

But whether we go through a deprecation or not, we still need to agree on what we think the new / future behaviour should be. 
I would say that it should follow what we already do for timedelta, so include it by default for dataframe.

**Comment by TomAugspurger** [comment]
Split the tests and fixed the merge conflicts.

I'm happy with this as is since it restores the default 1.0 behavior. If we want additional changes then let's do them as followups.

**Comment by jorisvandenbossche** [comment]
> I'm happy with this as is since it restores the default 1.0 behavior. If we want additional changes then let's do them as followups.

So I suppose this means you didn't change the dataframe behaviour (which is certainly fine to leave for a follow-up). But should be then maybe raise a NotImplementedError when specifying `datetime_as_numeric=True` for DataFrame?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. @jorisvandenbossche comment here or in a followup ok

**Comment by TomAugspurger** [comment]
@jorisvandenbossche can you remind me what the dataframe issue is? That datetimes are not included in this output while timedeltas are?

```python
In [4]: pd.DataFrame({'a': pd.date_range("2012", periods=3), 'b': [1, 2, 3]}).describe()
   ...:
Out[4]:
         b
count  3.0
mean   2.0
std    1.0
min    1.0
25%    1.5
50%    2.0
75%    2.5
max    3.0

In [6]: pd.DataFrame({'a': pd.timedelta_range("2012", periods=3), 'b': [1, 2, 3]}).describe()
   ...:
Out[6]:
                               a    b
count                          3  3.0
mean   1 days 00:00:00.000002012  2.0
std              1 days 00:00:00  1.0
min    0 days 00:00:00.000002012  1.0
25%    0 days 12:00:00.000002012  1.5
50%    1 days 00:00:00.000002012  2.0
75%    1 days 12:00:00.000002012  2.5
max    2 days 00:00:00.000002012  3.0
```

Interestingly if you have datetime, numeric, and timedelta, then the timedelta is not included:

```python
In [5]: pd.DataFrame({'a': pd.date_range("2012", periods=3), 'b': [1, 2, 3], 'c': pd.period_range('2000', periods=3)}).describe()
   ...:
Out[5]:
         b
count  3.0
mean   2.0
std    1.0
min    1.0
25%    1.5
50%    2.0
75%    2.5
max    3.0
```

This all seems buggy, but I hope can be handled separately.

**Comment by jorisvandenbossche** [comment]
Yes, that needs to be cleaned up and can indeed be handled separately. 

But the one thing that might be relevant for this PR:

```
pd.DataFrame({'a': pd.date_range("2012", periods=3), 'b': [1, 2, 3]}).describe(datetime_is_numeric=True)
```

is not actually doing what the keyword is expected to be doing (I suppose, didn't try with this branch). 
And if we want to enable that in a follow-up PR, we should maybe not allow one to pass that now for the DataFrame case (because if we would then "fix" it in 1.2, it's kind of a breaking change, although nobody of course should rely on it). So we could raise a NotImplementedError that keyword is specified and `self` is a DataFrame.

**Comment by TomAugspurger** [comment]
@jorisvandenbossche gotcha. Added a test and implemented that.

```python
In [3]: pd.DataFrame({'a': pd.date_range("2012", periods=3), 'b': [1, 2, 3]}).describe(datetime_is_numeric=True)
Out[3]:
                         a    b
count                    3  3.0
mean   2012-01-02 00:00:00  2.0
min    2012-01-01 00:00:00  1.0
25%    2012-01-01 12:00:00  1.5
50%    2012-01-02 00:00:00  2.0
75%    2012-01-02 12:00:00  2.5
max    2012-01-03 00:00:00  3.0
std                    NaN  1.0
```

**Comment by jreback** [comment]
hopefully fixed the conflict correctly. merging on green.

**Comment by TomAugspurger** [comment]
All green.

**Comment by TomAugspurger** [comment]
Err, wait, the whatsnew doesn't look right. Will fix.

**Comment by jreback** [comment]
cool

**Pull request merged**